### ISSUE TITLE
incorrect mepc

### ISSUE THREAD

**Issue by pentin-as** [riscv-boom/riscv-boom#39]
When executing the "ebreak" instruction, the mepc register is set to "PC + 4" instead of "PC".

**Comment by ccelio** [comment]
Verified to pass on rocket, fail on boom.

**Comment by pentin-as** [comment]
![break](https://user-images.githubusercontent.com/30556064/32259500-c6dc886c-bee2-11e7-94ca-22625212eb49.png)
![illegalinst](https://user-images.githubusercontent.com/30556064/32259505-ccca8f30-bee2-11e7-8232-38e7a722ba77.png)

When executing an illegal instruction (0x00000000) csr.io.pc is set in the correct state until its value is set in the epc register.
When executing the break instruction csr.io.pc is set in the correct state only one tact, the next value csr.io.pc is set in the epc register.

csr.io.pc is assigned from rob.io.com_xcpt.bits.pc
I think the error should be fixed in rob.io.com_xcpt.bits.pc

**Comment by ccelio** [comment]
What PC is your illegal instruction at? What value are you expecting?

**Comment by ccelio** [comment]
I think this is the error:

https://github.com/ucb-bar/riscv-boom/blob/master/src/main/scala/rob.scala#L645

It assumes that any `com_xcpt.bits.pc` that is not an exception wants PC+4.

**Comment by pentin-as** [comment]
PC with illegal instruction / "ebreak" - 0x1000000C.

**Comment by pentin-as** [comment]
I understand correctly that need a line:
 val refetch_inst = exception_thrown
change to:
 val refetch_inst = exception_thrown | insn_break
(and insn_break should be taken from the decoder)?
Or is it better to try completely to delete the expression "+ Mux(refetch_inst, UInt(0), UInt(4))" and run the tests?

I'll try the second method.

**Comment by pentin-as** [comment]
The second method did not help. I will understand how to do it right.

**Comment by ccelio** [comment]
Okay, here's how `ebreak` works in BOOM currently.

`ebreak` is a system instruction that is sent to the CSRFile for handling (it goes through the issue window, down the pipeline, and hits the CSRFile during "write back"). It is not an exception, so handling it in the ROB doesn't quite feel right. Trying to set some `insn_break` signal is non-obvious, because commit is looking across a row of potential instructions, some of which may not be valid. It would have to look at the first valid instruction across the potential commit row to see if it is a breakpoint (which can be set by decode).

I'm going to see if I can instead implement `ebreak` by getting its PC in the same manner as a branch instruction and feeding it directly to the CSR file once the `ebreak` flows down the pipeline to set the `csr.io.cmd`.

**Comment by ccelio** [comment]
Fixed by https://github.com/ucb-bar/riscv-boom/commit/71146913dc041362490893b413bae38dc10fd5f7. 

This issue will be closed once `devel` is merged into master.
